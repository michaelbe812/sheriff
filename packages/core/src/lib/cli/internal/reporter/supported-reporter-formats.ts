export type ReporterFormat = 'json' | 'junit';

export const SUPPORTED_REPORTER_FORMATS = ['json', 'junit'];
