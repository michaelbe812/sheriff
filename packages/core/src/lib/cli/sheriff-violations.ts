import { DependencyRuleViolation } from '../checks/check-for-dependency-rule-violation';

export type SheriffViolations = {
  encapsulationsCount: number;
  encapsulationValidations: string[];
  dependencyRuleViolationsCount: number;
  dependencyRuleViolations: DependencyRuleViolation[];
};
