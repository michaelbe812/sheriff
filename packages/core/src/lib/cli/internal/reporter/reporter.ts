import { SheriffViolations } from '../../sheriff-violations';

export interface Reporter {
  createReport(args: {
    exportDir: string;
    projectName: string;
    validationResults: SheriffViolations;
  }): void;
}
