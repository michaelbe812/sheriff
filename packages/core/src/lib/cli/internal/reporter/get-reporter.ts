import { JsonReporter } from './json-reporter';

export function getReporter(reporterFormat: string) {
  if (reporterFormat === 'json') {
    return new JsonReporter();
  }

  return null;
}
