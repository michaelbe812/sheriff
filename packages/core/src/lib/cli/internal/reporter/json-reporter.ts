import { Reporter } from './reporter';
import { cli } from '../../cli';
import getFs from '../../../fs/getFs';

import { SheriffViolations } from '../../sheriff-violations';

export class JsonReporter implements Reporter {
  createReport(args: {
    exportDir: string;
    projectName: string;
    validationResults: SheriffViolations;
  }): void {
    const fs = getFs();
    const targetPath = fs.join(
      args.exportDir,
      args.projectName,
      'sheriff-report' + this.#getReportExtension(),
    );
    cli.log(`Creating .json-export`);

    fs.createDir(fs.join(args.exportDir, args.projectName!));

    fs.writeFile(targetPath, JSON.stringify(args.validationResults, null, 2));
  }

  #getReportExtension(): string {
    return '.json';
  }
}
