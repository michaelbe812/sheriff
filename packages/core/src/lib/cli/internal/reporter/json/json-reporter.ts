import { Reporter } from '../reporter';
import { cli } from '../../../cli';
import getFs from '../../../../fs/getFs';

import { SheriffViolations } from '../../../sheriff-violations';

export class JsonReporter implements Reporter {
  #options: { outputDir: string; projectName: string };
  constructor(options: { outputDir: string; projectName: string }) {
    this.#options = options;
  }

  createReport(validationResults: SheriffViolations): void {
    const fs = getFs();
    const targetPath = fs.join(
      this.#options.outputDir,
      this.#options.projectName,
      'violations' + this.#getReportExtension(),
    );
    cli.log(`Creating JSON-export`);

    fs.createDir(fs.join(this.#options.outputDir, this.#options.projectName!));

    fs.writeFile(targetPath, JSON.stringify(validationResults, null, 2));
  }

  #getReportExtension(): string {
    return '.json';
  }
}
