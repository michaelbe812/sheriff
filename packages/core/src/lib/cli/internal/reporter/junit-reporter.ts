import { Reporter } from './reporter';

// TODO
//export class JunitReporter implements Reporter {}
