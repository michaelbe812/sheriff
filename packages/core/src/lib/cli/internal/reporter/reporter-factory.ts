import { JsonReporter } from './json-reporter';
import { Reporter } from './reporter';
import { JunitReporter } from './junit/junit-reporter';

export function reporterFactory(options: {
  reporterFormats: string[];
  outputDir: string;
  projectName: string;
}): Reporter[] {
  const reporters: Reporter[] = [];
  options.reporterFormats.forEach((format) => {
    if (format === 'json') {
      reporters.push(
        new JsonReporter({
          outputDir: options.outputDir,
          projectName: options.projectName,
        }),
      );
    }
    if (format === 'junit') {
      reporters.push(
        new JunitReporter({
          outputDir: options.outputDir,
          projectName: options.projectName,
        }),
      );
    }
  });

  return reporters;
}
