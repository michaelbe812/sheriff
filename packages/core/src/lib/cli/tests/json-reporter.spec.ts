import { beforeEach, describe, it, vitest, expect, beforeAll } from 'vitest';
import { VirtualFs } from '../../fs/virtual-fs';
import getFs, { useVirtualFs } from '../../fs/getFs';
import { JsonReporter } from '../internal/reporter/json-reporter';
import { toFsPath } from '../../file-info/fs-path';

import { SheriffViolations } from '../sheriff-violations';

describe('json reporter', () => {
  let fs: VirtualFs;
  beforeAll(() => {
    useVirtualFs();
    fs = getFs() as VirtualFs;
  });

  beforeEach(() => {
    fs.reset();
    fs.createDir('/project/customers/feature');
    fs.createDir('/project/shared/master-data');
    fs.createDir('/project/customers/ui');
    fs.createDir('/project/app/shared/form');
  });
  it('should create a json-file in /.sheriff/project/sheriff-report.json', () => {
    const reporter = new JsonReporter();
    const violations: SheriffViolations = {
      encapsulationsCount: 0,
      encapsulationValidations: [],
      dependencyRuleViolationsCount: 2,
      dependencyRuleViolations: [
        {
          rawImport: '@eternal/shared/master-data',
          fromModulePath: toFsPath('/project/customers/feature'),
          toModulePath: toFsPath('/project/shared/master-data'),
          fromTag: 'domain:customers',
          toTags: ['shared:master-data'],
        },
        {
          rawImport: '@eternal/shared/form',
          fromModulePath: toFsPath('/project/customers/ui'),
          toModulePath: toFsPath('/project/app/shared/form'),
          fromTag: 'domain:customers',
          toTags: ['shared:form'],
        },
      ],
    };
    reporter.createReport({
      exportDir: '.sheriff',
      projectName: 'project',
      validationResults: violations,
    });

    expect(
      fs.readFile('.sheriff/project/sheriff-report.json'),
    ).toMatchSnapshot();
  });
});
