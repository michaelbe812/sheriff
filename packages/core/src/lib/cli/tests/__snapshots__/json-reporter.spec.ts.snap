// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json reporter > should create a json-file in /.sheriff/project/sheriff-report.json 1`] = `
"{
  "encapsulationsCount": 0,
  "encapsulationValidations": [],
  "dependencyRuleViolationsCount": 2,
  "dependencyRuleViolations": [
    {
      "rawImport": "@eternal/shared/master-data",
      "fromModulePath": "/project/customers/feature",
      "toModulePath": "/project/shared/master-data",
      "fromTag": "domain:customers",
      "toTags": [
        "shared:master-data"
      ]
    },
    {
      "rawImport": "@eternal/shared/form",
      "fromModulePath": "/project/customers/ui",
      "toModulePath": "/project/app/shared/form",
      "fromTag": "domain:customers",
      "toTags": [
        "shared:form"
      ]
    }
  ]
}"
`;
